lib/FilesHooks.php

```diff
@@ -1219,4 +1210,154 @@ protected function commitActivityTransaction(bool $shouldFlush): void {
+		} catch (\Exception $e) {
+			return []; // if we have no access to RuleManager, we cannot filter unrelated users
+		}
+
+		/** @var \OCA\GroupFolders\ACL\Rule[] $rules */
+		$rules = $knownRules = $usersToCheck = [];
+		foreach ($cachedMounts as $cachedMount) {
+			// we are only interested in filtering GroupFolders ACL
+			if ($cachedMount['provider'] !== 'OCA\GroupFolders\Mount\MountProvider') {
+				continue;
+			}
+
+			// caching rules based on storage id
+			$storageId = $cachedMount['storageId'];
+			if (!array_key_exists($storageId, $knownRules)) {
+				$knownRules[$storageId] = [];
+			}
+
+			// caching rules based on storage+path to file
+			if (!array_key_exists($cachedMount['visiblePath'], $knownRules[$storageId])) {
+				$path = $cachedMount['path'];
+
+				// we need mountPoint and folderId to generate the correct path
+				try {
+					$node = $this->rootFolder->get($path);
+					$mountPoint = $node->getMountPoint();
+					$folderId = $mountPoint->getFolderId();
+					$path = substr($path, strlen($mountPoint->getMountPoint()));
+				} catch (\Exception $e) {
+					// in case of issue during the process, we can imagine the user have no access to the file
+					$usersToCheck[] = $cachedMount['userId'];
+					continue; // we'll catch rules on next user with access to the file
+				}
+
+				// we generate a list of path from top level of group folder to the file itself to get all rules
+				$paths = ['__groupfolders/' . $folderId];
+				while ($path !== '') {
+					$paths[] = '__groupfolders/' . $folderId . '/' . $path;
+					$path = dirname($path);
+					if ($path === '.' || $path === '/') {
+						$path = '';
+					}
+				}
+
+				// we might already know the rules for some path of the list
+				$paths = array_filter($paths, function (string $path) use ($knownRules, $storageId): bool {
+					if (array_key_exists($path, $knownRules[$storageId])) {
+						return false;
+					}
+
+					return true;
+				});
+
+				// we get and store the rules for each path from the list
+				$rulesPerPath = $ruleManager->getAllRulesForPaths($storageId, $paths);
+				foreach (array_keys($rulesPerPath) as $path) {
+					$rules = array_merge($rules, $rulesPerPath[$path]);
+				}
+
+				$knownRules[$storageId][$cachedMount['visiblePath']] = true;
+			}
+		}
+
+		// using each rules that disable read permission to generate a list of users
+		// that might not have access to fileId
+		foreach ($rules as $rule) {
+			if (($rule->getMask() & 1) === 0
+				|| ($rule->getPermissions() & 1) !== 0) {
+				continue; // not interested of rules with 'mask' not including read capability (1), or if 'permission' does
+			}
+
+			$mapping = $rule->getUserMapping();
+			$id = $mapping->getId();
+
+			// if mapping is about user
+			if ($mapping->getType() === 'user' && !in_array($id, $usersToCheck)) {
+				$usersToCheck[] = $id;
+			}
+
+			// if mapping is about group
+			if ($mapping->getType() === 'group') {
```

Comment:
Should keep a record of which groups we looped over already to avoid getting their user list more than once?

Reply:
this is improved in last commit